test/functional/pdo_sqlsrv/pdo_AE_functions.inc

```diff
@@ -0,0 +1,532 @@
+                }
+                ++$i;
+            }
+        }
+    }
+}
+
+// testCompare selects based on a comparison in the WHERE clause and compares
+// the results between encrypted and non-encrypted columns, checking that the
+// results are identical
+function TestCompare($conn, $tableName, $comparisons, $dataTypes, $colNames, $thresholds, $length, $key, $encryptionType, $attestation)
+{
+    foreach ($comparisons as $comparison) {
+        foreach ($dataTypes as $type) {
+
+            // Unicode data types require the PHPTYPE to be specified to UTF-8 string
+            // and the correct collation. Otherwise we get different results between
+            // the encrypted and non-encrypted columns (probably because the collation
+            // was only changed in the encryption query).
+            $string = dataTypeIsStringMax($type);
+            $collate = $string ? " COLLATE Latin1_General_BIN2" : "";
+            $unicode = dataTypeIsUnicode($type);
+            $PDOType = getPDOType($type);
+
+            $AEQuery = "SELECT ".$colNames[$type]."_AE FROM $tableName WHERE ".$colNames[$type]."_AE ".$comparison." ?".$collate;
+            $nonAEQuery = "SELECT ".$colNames[$type]." FROM $tableName WHERE ".$colNames[$type]." ".$comparison." ?".$collate;
+
+            try {
+                $AEstmt = $conn->prepare($AEQuery);
+                $AEstmt->bindParam(1, $thresholds[$type], $PDOType);
+                $nonAEstmt = $conn->prepare($nonAEQuery);
+                $nonAEstmt->bindParam(1, $thresholds[$type], $PDOType);
+            } catch (PDOException $error) {
+                print_r($error);
+                die("Preparing/binding statements for comparison failed");
+            }
+
+            compareResults($AEstmt, $nonAEstmt, $key, $encryptionType, $attestation, $comparison);
+        }
+    }
+}
+
+// testPatternMatch selects based on a pattern in the WHERE clause and compares
+// the results between encrypted and non-encrypted columns, checking that the
+// results are identical
+function TestPatternMatch($conn, $tableName, $patterns, $dataTypes, $colNames, $length, $key, $encryptionType, $attestation)
+{
+    foreach ($dataTypes as $type) {
+
+        // Pattern matching doesn't work in AE for non-string types
+        // without an explicit cast
+        if (!dataTypeIsStringMax($type)) {
+            continue;
+        }
+
+        foreach ($patterns[$type] as $pattern) {
+
+            $patternArray = array($pattern,
+                                  $pattern."%",
+                                  "%".$pattern,
+                                  "%".$pattern."%",
+                                  );
+
+            foreach ($patternArray as $spattern) {
+
+                $unicode = dataTypeIsUnicode($type);
+                $collate = $unicode ? " COLLATE Latin1_General_BIN2" : "";
+                $PDOType = getPDOType($type);
```

Comment:
This COLLATE clause is required for unicode comparisons only? Is it always the case or something specific for AE related comparisons?

Reply:
For AE comparisons in Unicode, a Latin1_General_BIN2 collation is required. I don't think it's required for ASCII data.